File: Two_Pointers/003.Longest-Substring-Without-Repeating-Characters/003.Longest-Substring-Without-Repeating-Characters.ts
function lengthOfLongestSubstring(s: string): number {
  const map = new Map<string, number>();
  let ret = 0;
  let start = 0;
  let end = 0;
  while (end < s.length) {
    if (!map.has(s[end]) || map.get(s[end]) === 0) {
      map.set(s[end], (map.get(s[end]) || 0) + 1);
      end++;
    } else {
      map.set(s[start], map.get(s[start]) - 1);
      start++;
    }
    ret = Math.max(ret, end - start);
  }

  return ret;
}

File: Two_Pointers/003.Longest-Substring-Without-Repeating-Characters/003.Longest-Substring-Without-Repeating-Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        notebook = {}
        result = 0
        
        start = 0
        for end in range(len(s)):
            notebook[s[end]] = notebook.setdefault(s[end], 0) + 1
            while notebook[s[end]] > 1:
                notebook[s[start]] = notebook.setdefault(s[start], 0) - 1
                start += 1
            result = max(result, end - start + 1)
        return result
